packages/@react-spectrum/s2/src/CardView.tsx

```diff
@@ -0,0 +1,511 @@
+/*
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {
+  GridList as AriaGridList,
+  GridLayoutOptions,
+  GridListItem,
+  GridListProps,
+  UNSTABLE_Virtualizer
+} from 'react-aria-components';
+import {CardContext, CardViewContext} from './Card';
+import {ImageCoordinator} from './ImageCoordinator';
+import {InvalidationContext, Layout, LayoutInfo, Rect, Size} from '@react-stately/virtualizer';
+import {Key, LoadingState, Node} from '@react-types/shared';
+import {style} from '../style/spectrum-theme' with {type: 'macro'};
+import {useLoadMore} from '@react-aria/utils';
+import {useMemo, useRef} from 'react';
+
+export interface CardViewProps<T> extends Omit<GridListProps<T>, 'layout' | 'keyboardNavigationBehavior' | 'selectionBehavior'> {
+  layout?: 'grid' | 'waterfall',
+  size?: 'XS' | 'S' | 'M' | 'L' | 'XL',
+  density?: 'compact' | 'regular' | 'spacious',
+  variant?: 'primary' | 'secondary' | 'tertiary' | 'quiet',
+  selectionStyle?: 'checkbox' | 'highlight',
+  loadingState?: LoadingState,
+  onLoadMore?: () => void
+}
+
+class FlexibleGridLayout<T extends object, O> extends Layout<Node<T>, O> {
```

Comment:
This is different from the GridLayout in RAC because it supports variable row heights - the height of a row is equal to the maximum height of the items in that row. The RAC GridLayout just scales the sizes of its items proportionally, which won't work for wrapping text. I could see potentially adding this as an option in RAC at some point.